mlserver/types/tracepoints.py

```diff
@@ -0,0 +1,186 @@
+
+    def all():
+        return {t for t in Tracepoint}
+
+    def none():
+        return set()
+
+    def get_arg_types(self) -> tuple[list[int] | None, ArgStatus]:
+        """
+        Returns a (arg_types_list, status) tuple for the current tracepoint. The
+        arg_types_list, when not None, defines the function prototype of the
+        tracepoint.
+
+        The returned arg_types_list is used by the TracingProvider when
+        registering the tracepoints with the underlying native tracepoint
+        implementation (libstapsdt). Based on this, libstapsdt dynamically
+        generates a shared library in ELF format, with one function
+        corresponding to each tracepoint and an ELF note section describing
+        where the tracepoint arguments can be found (in which registers) at call
+        time.
+
+        Tracepoints may have a maximum of MAX_TRACEPOINT_ARGS arguments. If the
+        defined prototype has more, or if the prototype can not be found, this
+        will be reflected in the returned ArgStatus. Any ArgStatus different
+        from ArgStatus.Ok indicates an error state, albeit some states such as
+        ArgStatus.TooManyArguments are recoverable (for example, by ignoring the
+        additional arguments).
+
+        The list of arguments used by each tracepoint is currently unstable.
+        """
+        if self in ArgTypeMap._prototypes:
+            if len(ArgTypeMap._prototypes[self]) <= MAX_TRACEPOINT_ARGS:
+                return ArgTypeMap._prototypes[self], ArgStatus.Ok
+            else:
+                return (
+                    ArgTypeMap._prototypes[self][0:MAX_TRACEPOINT_ARGS],
+                    ArgStatus.TooManyArguments,
+                )
+        else:
+            return None, ArgStatus.NoPrototypeDefined
+
+
+ArgTypeMap.model_args = [ArgTypeMap.model_name_t, ArgTypeMap.model_version_t]
+ArgTypeMap.model_args_reload = [
+    ArgTypeMap.model_name_t,
+    ArgTypeMap.model_version_t,
+    ArgTypeMap.old_model_version_t,
+]
+ArgTypeMap.queue_args = [ArgTypeMap.model_name_t, ArgTypeMap.queue_len_t]
+ArgTypeMap.infer_args = [
+    ArgTypeMap.model_name_t,
+    ArgTypeMap.pipeline_t,
+    ArgTypeMap.tags_t,
+    ArgTypeMap.worker_pid_t,
+]
+
+# TODO(lucian): Stabilise the tracepoint argument definitions after writing a
+# number of external applications using them
+#
+# Together with the tracepoint names, the prototypes form the public interface
+# against which external probing code is developed.
+ArgTypeMap._prototypes = {
+    Tracepoint.model_load_begin: ArgTypeMap.model_args,
+    Tracepoint.model_load_end: ArgTypeMap.model_args,
+    Tracepoint.model_reload_begin: ArgTypeMap.model_args_reload,
+    Tracepoint.model_reload_end: ArgTypeMap.model_args,
+    Tracepoint.model_unload: ArgTypeMap.model_args,
+    Tracepoint.inference_enqueue_req: ArgTypeMap.queue_args,
+    Tracepoint.inference_begin: ArgTypeMap.infer_args,
+    Tracepoint.inference_end: ArgTypeMap.infer_args,
+}
```

Comment:
Thinking out loud here, but the tracepoints and args seem very coupled with the implementation - would it be possible to make those more generic? 

My main worry would be whether a small change into any of those implementations (e.g. changing the args for inference) could then cause any side effects around system tracing.

Reply:
In a way, they do end quite coupled indeed. However, the prototypes essentially become the public interface that external probes use to run their code. Any changes to arguments, besides simply adding new ones at the end of the list, would require changes for whatever external BPF probes people write against the interface. So we would need to stabilise them at some point, after we write more actual probing code.

The idea is to pass to probes the minimal information that is _required_ for meaningful tracing applications. Hopefully, this information remains relatively stable even as MLServer evolves. For example, the actual inference functions might change without us changing the data passed for tracing via the same Tracepoints.

That being said, the `_prototypes` defined here do not end up being "enforced". Say we add a new argument to the `inference_begin` tracepoint without changing any of the `tp_*` functions in the provider. All existing code will continue to work because the function for triggering the tracepoint (`Probe.fire`) receives a variable number of arguments. We can then selectively call the tracepoint with additional arguments _sometimes_ by directly using the `SystemTracingProvider.__call__(tracepoint, *args)` (I don't necessarily advocate that it's a sane thing to do, as code might end up being confusing in the long term wrt to what arguments should be passed to which tracepoints).

We can discuss further, perhaps there is a better design to be found here.

Reply:
Thanks for adding that extra context. 

Do you have an example of how it looks on the "other end", i.e. where those probes are used?